planner/core/rule_partition_processor.go

```diff
@@ -224,13 +224,80 @@ func (s *partitionProcessor) processHashPartition(ds *DataSource, pi *model.Part
 	return tableDual, nil
 }
 
-func (s *partitionProcessor) pruneListPartition(tbl table.Table, partitionNames []model.CIStr) []int {
-	if len(partitionNames) > 0 {
-		pi := tbl.Meta().Partition
+func (s *partitionProcessor) findUsedListPartitions(ctx sessionctx.Context, tbl table.Table, partitionNames []model.CIStr,
+	conds []expression.Expression) ([]int, error) {
+	pi := tbl.Meta().Partition
+	partExpr, err := tbl.(partitionTable).PartitionExpr()
+	if err != nil {
+		return nil, err
+	}
+	pruneList := partExpr.ForListPruning
+	cols := make([]*expression.Column, 0, len(pruneList.ExprCols))
+	colLen := make([]int, 0, len(pruneList.ExprCols))
+	for _, c := range pruneList.ExprCols {
+		cols = append(cols, c.Clone().(*expression.Column))
+		colLen = append(colLen, types.UnspecifiedLength)
+	}
+
+	datchedResult, err := ranger.DetachCondAndBuildRangeForPartition(ctx, conds, cols, colLen)
+	if err != nil {
+		return nil, err
+	}
+
+	ranges := datchedResult.Ranges
+	used := make([]int, 0, len(ranges))
+	for _, r := range ranges {
+		if r.IsPointNullable(ctx.GetSessionVars().StmtCtx) {
+			if !r.HighVal[0].IsNull() {
+				if len(r.HighVal) != len(cols) {
+					used = []int{-1}
+					break
+				}
+			}
+			found := int64(-1)
+			for j, expr := range pruneList.Exprs {
+				ret, _, err := expr.EvalInt(ctx, chunk.MutRowFromDatums(r.HighVal).ToRow())
+				if err != nil {
+					return nil, err
+				}
+				if ret > 0 {
+					found = int64(j)
+					break
+				}
+			}
+			if found == -1 {
+				continue
+			}
+			if len(partitionNames) > 0 && !s.findByName(partitionNames, pi.Definitions[found].Name.L) {
+				continue
+			}
+			used = append(used, int(found))
+		} else {
+			used = []int{FullRange}
+			break
+		}
+	}
+	if len(partitionNames) > 0 && len(used) == 1 && used[0] == FullRange {
 		or := partitionRangeOR{partitionRange{0, len(pi.Definitions)}}
-		return s.convertToIntSlice(or, pi, partitionNames)
+		return s.convertToIntSlice(or, pi, partitionNames), nil
+	}
+	sort.Ints(used)
```

Comment:
Why not make `used` a map as long as you need to deduplication here?

Reply:
Great, done.